### PR TITLE
AutoGluon Tabular HPO Optimization + Refactoring

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

Major Changes:
- Added fold-level fit control to bagged models
- Optimized HPO when bagging (speedup 10-20%)
- Optimized KNN memory usage during HPO (halved peak memory usage)
- Various code cleanup / refactoring to standardize model parameters

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-188-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-188/1/index.html

File: autogluon/task/tabular_prediction/tabular_prediction.py [hunk 1]
```diff
@@ -129,12 +129,16 @@ def fit(train_data, label, tuning_data=None, output_directory=None, problem_type
...
                      and the CatBoost docs: https://catboost.ai/docs/concepts/parameter-tuning.html
-                RF: `n_estimators` is currently the only hyperparameter you can specify, 
-                    see sklearn docs: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html 
-                XT: `n_estimators` is currently the only hyperparameter you can specify, 
-                    see sklearn docs: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.ExtraTreesClassifier.html
-                KNN: Currently no hyperparameters may be specified for k-nearest-neighbors models
-                
+                RF: See sklearn docs: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html
+                    Note: Hyperparameter tuning is disabled for this model.
+                    Note: 'criterion' parameter will be overriden.
```
Comment (jwmueller) [inline autogluon/task/tabular_prediction/tabular_prediction.py]: should clarify: 'criterion' parameter will be overridden by what?

File: autogluon/task/tabular_prediction/tabular_prediction.py [hunk 2]
```diff
@@ -129,12 +129,16 @@ def fit(train_data, label, tuning_data=None, output_directory=None, problem_type
...
+                XT: See sklearn docs: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.ExtraTreesClassifier.html
+                    Note: Hyperparameter tuning is disabled for this model.
+                    Note: 'criterion' parameter will be overriden.
```
Comment (jwmueller) [inline autogluon/task/tabular_prediction/tabular_prediction.py]: should clarify: 'criterion' parameter will be overridden by what?

File: autogluon/task/tabular_prediction/tabular_prediction.py [hunk 3]
```diff
@@ -129,12 +129,16 @@ def fit(train_data, label, tuning_data=None, output_directory=None, problem_type
...
+                KNN: See sklearn docs: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html
+                    Note: Hyperparameter tuning is disabled for this model.
+                    Note: 'weights' parameter will be overriden.
```
Comment (jwmueller) [inline autogluon/task/tabular_prediction/tabular_prediction.py]: should clarify: 'weights' parameter will be overridden by what?

File: autogluon/task/tabular_prediction/tabular_prediction.py (hunk 1 above)
Comment (Innixma) [reply autogluon/task/tabular_prediction/tabular_prediction.py]: Added

File: autogluon/task/tabular_prediction/tabular_prediction.py (hunk 2 above)
Comment (Innixma) [reply autogluon/task/tabular_prediction/tabular_prediction.py]: Added

File: autogluon/task/tabular_prediction/tabular_prediction.py (hunk 3 above)
Comment (Innixma) [reply autogluon/task/tabular_prediction/tabular_prediction.py]: Added

**Comment by szha** [comment]
Job PR-188-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-188/2/index.html

**Pull request merged**